src/lib/efhw/af_xdp.c

```diff
@@ -437,19 +442,50 @@ static int xdp_bind(struct socket* sock, int ifindex, unsigned queue, unsigned f
 }
 
 /* Link an XDP program to an interface */
-static int xdp_set_link(struct net_device* dev, struct bpf_prog* prog)
-{
-  struct netdev_bpf bpf = {
-    .command = XDP_SETUP_PROG,
-    .prog = prog
-  };
-
-  if( !dev->netdev_ops->ndo_bpf ) {
-    EFHW_ERR("%s: %s does not support XDP", __FUNCTION__, dev->name);
-    return -ENOSYS;
+static int xdp_set_link(struct net_device* dev, int prog_fd)
+{
+  if( dev->netdev_ops->ndo_bpf ) {
+    struct netdev_bpf bpf = {
+      .command = XDP_SETUP_PROG,
+      .prog = NULL,
+    };
+
+    if( prog_fd > 0 ) {
+      struct bpf_prog* prog = bpf_prog_get_type_dev(prog_fd, BPF_PROG_TYPE_XDP, 1);
+      ci_close_fd(prog_fd);
+      if( IS_ERR(prog) )
+        return PTR_ERR(prog);
+      bpf.prog = prog;
+    }
+
+    return dev->netdev_ops->ndo_bpf(dev, &bpf);
+  }
+  else {
+    char *envp[] = { NULL };
+    char *argv[] = {
+      NULL,
+      dev->name,
+      prog_fd > 0 ? XDP_PROG_NAME : NULL,
+      NULL
+    };
+
+    EFHW_WARN("%s: %s does not support native XDP, using generic mode",
+              __FUNCTION__, dev->name);
+
+    if( bpf_link_helper ) {
+      argv[0] = strim(bpf_link_helper);
+    }
+    else {
+      EFHW_ERR("%s: bpf_link_helper parameter is not set. Failed to link.",
+               __FUNCTION__);
+      return -1;
+    }
+
+    EFHW_WARN("%s: spawning %s %s %s", __FUNCTION__,
+              argv[0], argv[1], argv[2] ? argv[2] : "");
+
+    return call_usermodehelper(argv[0], argv, envp, UMH_WAIT_EXEC);
```

Comment:
Let me shortlist the patch for review and testing. While enabling generic XDP is an improvement over our current constraints, we aren't entirely sure if we want to go with the "synchronous" approach of asking the userland to link a program. Instead, we may let the users link whenever they prefer after starting an onloaded `AF_XDP` application, e.g. with https://github.com/xdp-project/xdp-tools/tree/master/xdp-loader.

Reply:
Oh, sorry for the confusion. What I meant is that long term, we aren't 100% committed to the approach of configuring BPF from the kernel space and may consider moving that bit to UL. This theoretically would remove the necessity of using `bpf-link-helper`, hence the comment.

On second thought, disregard it because I suppose it won't be much trouble to refactor the generic XDP if we create a map and loan a BPF program in UL.

Hope that does make sense!